### PR TITLE
Better handling of index deletion during snapshot

### LINKED ISSUE #9024: NPE in SnapshotsService in 1.4.2
I keep getting the same NPE when performing snapshots in 1.4.2:

```
java.lang.NullPointerException
	at org.elasticsearch.snapshots.SnapshotsService.shards(SnapshotsService.java:1195)
	at org.elasticsearch.snapshots.SnapshotsService.access$800(SnapshotsService.java:88)
	at org.elasticsearch.snapshots.SnapshotsService$2.execute(SnapshotsService.java:300)
	at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:329)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:153)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:744)
```

The line in question is:

`IndexRoutingTable indexRoutingTable = clusterState.getRoutingTable().index(index);`

clusterState isn't null because it's successfully dererenced above.  So that leaves getRoutingTable().  Any ideas as to why that would return null?

### PR DESCRIPTION
**Pull request by imotov**
If an index is deleted during initial state of the snapshot operation, the entire snapshot can fail with NPE. This commit improves handling of this situation and allows snapshot to continue if partial snapshots are allowed.

Closes #9024

### CODE REVIEW

File: src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreTests.java [hunk 1]
```diff
@@ -1558,6 +1558,58 @@ public void changeSettingsOnRestoreTest() throws Exception {
...
+        refresh();
+        assertThat(client.prepareCount("test-idx-1").get().getCount(), equalTo(100L));
+        assertThat(client.prepareCount("test-idx-2").get().getCount(), equalTo(100L));
+        assertThat(client.prepareCount("test-idx-3").get().getCount(), equalTo(100L));
+
+        logger.info("--> snapshot allow partial {}", allowPartial);
+        ListenableActionFuture<CreateSnapshotResponse> future = client.admin().cluster().prepareCreateSnapshot("test-repo", "test-snap")
+                .setIndices("test-idx-*").setWaitForCompletion(true).setPartial(allowPartial).execute();
+        logger.info("--> delete some indices while snapshot is running");
+        client.admin().indices().prepareDelete("test-idx-1", "test-idx-2").get();
```
Comment (rjernst) [inline src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreTests.java]: How can you be sure the snapshot didn't complete prior to running these deletes without some type of latch?

**Comment by rjernst** [comment]
The fix looks good to me, but I'm concerned the test could be flaky with timing issues.

**Comment by imotov** [comment]
@rjernst Good point. I made test more deterministic. Could you take a look?

**Comment by rjernst** [comment]
Thanks @imotov, LGTM.

**Pull request merged**